middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -88,27 +88,27 @@ let check_alloc_fields = function
 let make_block ~dbg kind alloc_mode args =
   check_alloc_fields args;
   let mode = Alloc_mode.For_allocations.to_lambda alloc_mode in
-  let allocator, tag =
-    match (kind : P.Block_kind.t) with
-    | Values (tag, _) -> C.make_alloc, Tag.Scannable.to_tag tag
-    | Naked_floats -> C.make_float_alloc, Tag.double_array_tag
-    | Mixed (tag, shape) ->
-      let value_prefix_size = K.Mixed_block_shape.value_prefix_size shape in
-      let flat_suffix =
-        Array.map
-          (fun (flat_elt : K.Flat_suffix_element.t) : C.Flat_suffix_element.t ->
-            match flat_elt with
-            | Tagged_immediate -> Tagged_immediate
-            | Naked_float -> Naked_float
-            | Naked_float32 -> Naked_float32
-            | Naked_int32 -> Naked_int32
-            | Naked_int64 | Naked_nativeint -> Naked_int64_or_nativeint)
-          (K.Mixed_block_shape.flat_suffix shape)
-      in
-      ( C.make_mixed_alloc ~value_prefix_size ~flat_suffix,
-        Tag.Scannable.to_tag tag )
-  in
-  allocator ~mode dbg ~tag:(Tag.to_int tag) args
+  match (kind : P.Block_kind.t) with
+  | Values (tag, _) ->
+    let tag = Tag.Scannable.to_int tag in
+    C.make_alloc ~mode dbg ~tag args
+  | Naked_floats ->
+    let tag = Tag.to_int Tag.double_array_tag in
+    C.make_float_alloc ~mode dbg ~tag args
+  | Mixed (tag, shape) ->
+    let value_prefix_size = K.Mixed_block_shape.value_prefix_size shape in
+    let args_memory_chunks =
+      Array.fold_right
+        (fun kind acc ->
+          let chunk =
+            To_cmm_shared.memory_chunk_of_kind (K.With_subkind.anything kind)
+          in
+          chunk :: acc)
+        (K.Mixed_block_shape.field_kinds shape)
+        []
+    in
+    let tag = Tag.Scannable.to_int tag in
+    C.make_mixed_alloc ~mode dbg ~tag ~value_prefix_size args args_memory_chunks
```

Comment:
It's not clear to me why this diff is needed to actually solve the problem at hand (see other comment about `Flat_suffix_element` too).

Reply:
This part of the diff actually helps ensure that we do not make any error with indexes.

In the new code, allocating a block requires a list of expression, along with the corresponding `memory_chunks`. This gives **all** the information needed to generate the correct writes, especially it gives a very simple and correct way to compute the offset and size of each writes, all the while being simple to generate (because we have the kinds on the flambda side).

Without that change, there is a lot more code that need to compute offsets at various and different points in the code (with a lot of potential errors): if one only provides the kinds of the flat suffix, we then need to infer the size of all of the things before, and to do that, we need some assumptions which may turn out to change at one point, and we can't easily test when those assumptions stop being true (also, it duplicates the computation of indexes and offsets, which is superfluous and error prone, since one needs to make sure that multiple pieces of code compute indexes that are coherent).